examples/stm32/xively_demo_stm32f4xx_nucleo_wifi/Src/main.c

```diff
@@ -78,180 +80,298 @@
+#define DEBUG_UART_BAUDRATE 115200
+#define WIFI_BOARD_UART_BAUDRATE 115200
+#define USE_FLASH_STORAGE 1
 #define WIFI_SCAN_BUFFER_LIST 25
 
-/* Private function prototypes -----------------------------------------------*/
-WiFi_Status_t wifi_get_AP_settings( void );
+#define USE_HARDCODED_CREDENTIALS 0 /* Overrules USE_FLASH_STORAGE */
+#if USE_HARDCODED_CREDENTIALS
+#define USER_CONFIG_WIFI_SSID "User's WiFi Network Name"
+#define USER_CONFIG_WIFI_PWD "User's WiFi Network Password"
+#define USER_CONFIG_WIFI_ENCR WPA_Personal /* [ WPA_Personal | WEP | None ] */
+#define USER_CONFIG_XI_ACCOUNT_ID "Xively Account ID"
+#define USER_CONFIG_XI_DEVICE_ID "Xively Device ID"
+#define USER_CONFIG_XI_DEVICE_PWD "Xively Device Password"
+#endif
+/* Private macro -------------------------------------------------------------*/
+
+/* Private variables ---------------------------------------------------------*/
+static user_data_t user_config;
 
 wifi_state_t wifi_state;
-wifi_config config;
+wifi_config wifi_module_config;
 wifi_scan net_scan[WIFI_SCAN_BUFFER_LIST];
 xi_state_t xi_connection_state = XI_SOCKET_NO_ACTIVE_CONNECTION_ERROR;
-uint8_t is_burst_mode          = 0;
-
-char console_ssid[40];
-char console_psk[20];
+uint8_t virtual_switch_state   = 0; /* Pressing the user button flips this value */
+#define is_burst_mode virtual_switch_state
 
-char* ssid          = "SSID";
-char* seckey        = "PASSWORD";
-WiFi_Priv_Mode mode = WPA_Personal;
+xi_context_handle_t gXivelyContextHandle      = XI_INVALID_CONTEXT_HANDLE;
+xi_timed_task_handle_t gXivelyTimedTaskHandle = XI_INVALID_TIMED_TASK_HANDLE;
 
-#define XI_ACCOUNT_ID "XIVELY ACCOUNT ID"
-#define XI_DEVICE_ID "XIVELY DEVICE ID"
-#define XI_DEVICE_PASS "XIVELY DEVICE PASSWORD"
+/* Private function prototypes -----------------------------------------------*/
+static inline void print_user_config_debug_banner( void );
+static int8_t user_config_init( user_data_t* dst );
+static int8_t get_ap_credentials( user_data_t* udata );
+static int8_t get_xively_credentials( user_data_t* udata );
 
 /* the interval for the time function */
 #define XI_PUBLISH_INTERVAL_SEC 5
 
 /* combines name with the account and device id */
-#define XI_TOPIC_NAME_MANGLE( name )                                                     \
-    "xi/blue/v1/" XI_ACCOUNT_ID "/d/" XI_DEVICE_ID "/" name
+#define XI_TOPIC_MAX_LEN 256
+
+/* Xively channel names - These weill be used to build the MQTT topic addresses */
+#define ACCELEROMETER_CHANNEL_NAME "Accelerometer"
+#define GYROSCOPE_CHANNEL_NAME     "Gyroscope"
+#define MAGNETOMETER_CHANNEL_NAME  "Magnetometer"
+#define BAROMETER_CHANNEL_NAME     "Barometer"
+#define HUMIDITY_CHANNEL_NAME      "Humidity"
+#define TEMPERATURE_CHANNEL_NAME   "Temperature"
+#define BUTTON_CHANNEL_NAME        "Button"
+#define LED_CHANNEL_NAME           "LED"
+
+/* RAM can be saved by building each of these when needed and free()ing them after use */

```

Comment:
We're not, that's why we're not free()ing those strings. I just added the comment in case a user used the example as a template for a device with a smaller footprint, but I'm happy to remove it if you think it's confusing.